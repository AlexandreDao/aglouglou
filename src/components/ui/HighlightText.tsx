import { StyleSheet, Text, TextStyle, TextProps } from 'react-native'
import React from 'react'
import { ACTIVE_COLOR } from '@/constants/colors'
import { normalizeString } from '@/utils/stringUtils'

interface HighlightTextProps {
  textToHighlight?: string
  numberOfLines?: number
  children: string
  style?: TextStyle
}

const styles = StyleSheet.create({
  highlightedText: {
    color: ACTIVE_COLOR,
  },
})

const HighlightText = ({ textToHighlight, children, style, numberOfLines }: HighlightTextProps) => {
  if (!textToHighlight)
    return (
      <Text style={style} numberOfLines={numberOfLines}>
        {children}
      </Text>
    )

  const normalizedTextToHighlight = normalizeString(textToHighlight).toLocaleLowerCase()
  const regex = new RegExp(`(${normalizedTextToHighlight})`, 'gi')
  const parts = children.split(regex)
  const highlightIndex = parts.findIndex((p) => normalizeString(p).toLocaleLowerCase() === normalizedTextToHighlight)
  const ratio = highlightIndex / parts.length
  let ellipsizeMode: TextProps['ellipsizeMode']

  if (ratio > 0.6) {
    ellipsizeMode = 'tail'
  } else if (ratio < 0.4) {
    ellipsizeMode = 'head'
  } else {
    ellipsizeMode = 'middle'
  }

  return (
    <Text style={style} numberOfLines={numberOfLines} ellipsizeMode={ellipsizeMode}>
      {parts.map((part, index) =>
        index === highlightIndex ? (
          <Text key={index} style={styles.highlightedText}>
            {part}
          </Text>
        ) : (
          <Text key={`highlighttext-${index}`}>{part}</Text>
        )
      )}
    </Text>
  )
}

export default HighlightText
