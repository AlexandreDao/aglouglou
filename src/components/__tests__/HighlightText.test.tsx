import React from 'react'
import { render } from '@testing-library/react-native'
import HighlightText from '@/components/ui/HighlightText'

describe('<HighlightText />', () => {
  test('HighlightText snapshot', () => {
    const displayedString = 'Hello world'
    const tree = render(<HighlightText textToHighlight="Hello">{displayedString}</HighlightText>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('HighlightText text is highlight', () => {
    const displayedString = 'Hello world'
    const { getByText } = render(<HighlightText textToHighlight="Hello">{displayedString}</HighlightText>)

    expect(getByText('Hello')).toBeDefined()
    expect(getByText('Hello world')).toBeDefined()
  })
})
