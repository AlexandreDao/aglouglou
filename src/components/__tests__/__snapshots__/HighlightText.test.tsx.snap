// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HighlightText /> HighlightText snapshot 1`] = `
<Text
  ellipsizeMode="head"
>
  <Text />
  <Text
    style={
      {
        "color": "#FF8EB6",
      }
    }
  >
    Hello
  </Text>
  <Text>
     world
  </Text>
</Text>
`;
